blocks/bricks/wrappers.py

```diff
@@ -2,40 +2,91 @@
 from theano import tensor
 
 from blocks.bricks.base import Brick, application
+from blocks.utils import dict_subset, pack
 
 
-class As2D(Brick):
-    """Wraps a 2D application around reshapes to accomodate N-D tensors.
+class WithExtraDims(Brick):
```

Comment:
>  But in this particular situation I feel that wrapping the apply with another brick approach is not good - it is not intuitive and requires to instantiate softmax twice. 

I think I know how to refactor `WithExtraDims` and `WithAxesSwapped` to be metaclasses. This in many aspect will combine best of two worlds. 

Have you ever had to use `categorical_cross_entropy` with `axis` other than `-1`? I can imagine situation in which `softmax` has to be used with a different axis, but for categorical cross entropy, I can't.